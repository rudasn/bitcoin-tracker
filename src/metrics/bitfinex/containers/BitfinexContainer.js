import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { compose } from 'recompact'
import propTypes from 'prop-types'

import Panel from 'metrics/components/Panel'
import { withAPI } from 'api'

import * as actions from '../actions'
import getTickerOptions from '../tickers'

export class BitfinexContainer extends PureComponent {
  constructor(props) {
    super(props)
    this.fetch = this.fetch.bind(this)
    this.initialize = this.initialize.bind(this)
    this.teardown = this.teardown.bind(this)
    this.initialized = false
    this.fetchInterval = null
  }
  componentDidMount() {
    this.initialize()
  }
  componentWillUnmount() {
    this.teardown()
  }
  componentDidUpdate(prevProps) {
    if (prevProps.ticker !== this.props.ticker) {
      this.initialize()
    }
  }
  initialize() {
    const { updateEvery } = this.props
    if (this.initialized) {
      this.teardown()
    }
    this.fetch()
    this.fetchInterval = setInterval(() => {
      this.fetch()
    }, updateEvery)
    this.initialized = true
  }
  teardown() {
    clearInterval(this.fetchInterval)
  }
  fetch() {
    const { dispatch, ticker, api } = this.props
    dispatch(actions.fetchBitfinex({ ticker, api, }))
  }
  render() {
    const { title, points, property, lastChange, currency, error, loading } = this.props
    const { chartOptions, tableOptions } = getTickerOptions(this.props) // todo: memo

    const last = points[ points.length - 1 ] || {}
    const value = last.point ? `${ currency }${ last.point[property] }` : null
    const changes = lastChange[property] || [ 0, 0 ]
    const [ change, percentage ] = changes

    return (
      <Panel
        title={ title }
        value={ value }
        change={ change }
        error={ error }
        loading={ loading }
        dateUpdated={ last.date }
        percentage={ percentage }
        chartOptions={ chartOptions }
        tableOptions={ tableOptions }
        />
    )
  }
}

BitfinexContainer.propTypes = {
  currency: propTypes.string,
  property: propTypes.string.isRequired,
  ticker: propTypes.string.isRequired,
  title: propTypes.string,
  updateEvery: propTypes.number,

  // from API HOC
  api: propTypes.object.isRequired,

  // from Redux state
  error: propTypes.object,
  lastChange: propTypes.object,
  lastChecked: propTypes.instanceOf(Date),
  lastResponse: propTypes.instanceOf(Date),
  loading: propTypes.bool,
  points: propTypes.arrayOf(propTypes.object),
}

BitfinexContainer.defaultProps = {
  points: [],
  updateEvery: 30000,
  lastChange: {},
  currency: '',
}

const mapStateToProps = (state, props) => {
  const { ticker } = props
  return state.metrics.bitfinex[ticker] || {}
}

export default compose(
  withAPI,
  connect(mapStateToProps)
)(BitfinexContainer)
