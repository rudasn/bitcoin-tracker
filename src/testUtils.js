import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import date from 'utils/date'

export const mockStore = configureMockStore([ thunk ])
export const mockApi = (response, error) => {
  const id = Date.now()
  const _mockApiGet = jest.fn()
  return {
    _mockApiGet,
    get: () => {
      _mockApiGet(error ? error : response)
      return error ? Promise.reject(error) : Promise.resolve(response)
    },
  }
}
export const mockDate = date
