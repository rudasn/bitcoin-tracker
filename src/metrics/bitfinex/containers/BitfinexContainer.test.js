import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { mockStore, mockApi, mockDate } from 'testUtils'

import { BitfinexContainer } from './BitfinexContainer'

it('renders without crashing', () => {
  const api = mockApi({})
  const store = mockStore({ metrics: { bitfinex: {} } })
  const element = (
    <BitfinexContainer
      property="last_price"
      ticker="btcgbp"
      api={ api }
      dispatch={ store.dispatch }
    />
  )
  const div = document.createElement('div')
  ReactDOM.render(element, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders initial state', () => {
  const api = mockApi({})
  const store = mockStore({ metrics: { bitfinex: {} } })
  const element = (
    <BitfinexContainer
      property="last_price"
      ticker="btcgbp"
      api={ api }
      dispatch={ store.dispatch }
    />
  )
  const tree = renderer.create(element).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders success state', () => {
  const api = mockApi({})
  const store = mockStore({ metrics: { bitfinex: {} } })
  const element = (
    <BitfinexContainer
      property="last_price"
      ticker="btcgbp"
      currency="GBP"
      api={ api }
      dispatch={ store.dispatch }
      points={[
        { point: { last_price: "1234.456" }, date: mockDate() },
      ]}
    />
  )
  const tree = renderer.create(element).toJSON()
  expect(tree).toMatchSnapshot()
})

describe('refreshes after a certain amount of time', () => {
  const api = mockApi({})
  const store = mockStore({ metrics: { bitfinex: {} } })
  const element = (
    <BitfinexContainer
      property="last_price"
      ticker="btcgbp"
      currency="GBP"
      api={ api }
      dispatch={ store.dispatch }
      points={[
        { point: { last_price: "1234.456" }, date: mockDate() },
      ]}
      updateEvery={ 1 }
    />
  )

  jest.useFakeTimers()
  const div = document.createElement('div')
  ReactDOM.render(element, div)

  it('fetches data on mount', () => {
    expect(api._mockApiGet).toHaveBeenCalledTimes(1)
  })

  it('fetches data every interval', () => {
    jest.runTimersToTime(10)
    expect(api._mockApiGet).toHaveBeenCalledTimes(11)
  })

  it('stops fetching when component is unmounted', () => {
    ReactDOM.unmountComponentAtNode(div)
    jest.runTimersToTime(100)
    expect(api._mockApiGet).toHaveBeenCalledTimes(11)
  })
})
