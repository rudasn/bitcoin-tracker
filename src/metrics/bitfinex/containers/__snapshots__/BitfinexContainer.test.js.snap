// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial state 1`] = `
<section
  className="panel"
>
  <header
    className="panel-header"
  >
    <div
      className="panel-state"
    />
    <h2
      className="panel-title"
    />
    <h3
      className="panel-subtitle"
    >
      <div
        className="ticker"
      >
        <span
          className="change-colored-text"
        >
          <span
            className="arrow"
          />
        </span>
        <span>
           
        </span>
        <span
          className="change-colored-text"
        >
          0.00
           (
          0.00
          %)
        </span>
      </div>
    </h3>
  </header>
  <main
    className="panel-main"
  >
    <section
      className="panel-chart"
    >
      {
          "options": {
              "theme": "light2",
              "zoomEnabled": true,
              "axisY": {
                  "includeZero": false,
                  "interval": 5,
                  "prefix": ""
              },
              "axisX": {
                  "interval": 30,
                  "crosshair": {
                      "enabled": true
                  }
              },
              "data": [
                  {
                      "type": "area",
                      "lineThickness": 1,
                      "markerType": "circle",
                      "xValueFormatString": "HH:mm:ss.ff",
                      "yValueFormatString": "#,##0.00",
                      "dataPoints": []
                  }
              ]
          }
      }
    </section>
  </main>
  <footer
    className="panel-footer"
  >
    <p>
      <small>
        Last updated: 
        <strong>
          -
        </strong>
      </small>
    </p>
  </footer>
</section>
`;

exports[`renders success state 1`] = `
<section
  className="panel"
>
  <header
    className="panel-header"
  >
    <div
      className="panel-state"
    />
    <h2
      className="panel-title"
    />
    <h3
      className="panel-subtitle"
    >
      <div
        className="ticker"
      >
        <span
          className="change-colored-text"
        >
          <span
            className="arrow"
          />
        </span>
        <span>
          GBP1234.456
           
        </span>
        <span
          className="change-colored-text"
        >
          0.00
           (
          0.00
          %)
        </span>
      </div>
    </h3>
  </header>
  <main
    className="panel-main"
  >
    <section
      className="panel-chart"
    >
      {
          "options": {
              "theme": "light2",
              "zoomEnabled": true,
              "axisY": {
                  "includeZero": false,
                  "interval": 5,
                  "prefix": "GBP"
              },
              "axisX": {
                  "interval": 30,
                  "crosshair": {
                      "enabled": true
                  }
              },
              "data": [
                  {
                      "type": "area",
                      "lineThickness": 1,
                      "markerType": "circle",
                      "xValueFormatString": "HH:mm:ss.ff",
                      "yValueFormatString": "GBP#,##0.00",
                      "dataPoints": [
                          {
                              "x": "2018-01-01T00:00:00.000Z",
                              "y": 1234.456
                          }
                      ]
                  }
              ]
          }
      }
    </section>
  </main>
  <footer
    className="panel-footer"
  >
    <p>
      <small>
        Last updated: 
        <strong>
          Mon Jan 01 2018 02:00:00 GMT+0200 (Eastern European Standard Time)
        </strong>
      </small>
    </p>
  </footer>
</section>
`;
